File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> counts;
        for (int num : nums) {
            if (counts.contains(num))
                counts[num]++;
            else
                counts[num] = 1;
        }
        vector<int> result {};
        for (auto& p : counts) {
            if (p.second >= k)
                result.push_back(p.first);
        }
        return result;
    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        """
        min-heap
        """
        counts = defaultdict(int)
        for num in nums:
            counts[num] += 1

        heap = []
        for num, count in counts.items():
            heapq.heappush(heap, (count, num))
            if len(heap) > k:
                heapq.heappop(heap)

        res = []
        for _, num in heap:
            res.append(num)

        return res
